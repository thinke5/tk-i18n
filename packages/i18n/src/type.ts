export type StrAst = {
  type: string;
  value: string;
  toString: () => string;
  // state?: any;
  varName?: string;
};

export type Ast = StrAst[];

export type VarHandle = {
  type: string;
  /** 如何区分 */
  separate: (str: string) => null | StrAst;
};

export type Config = {
  language: string[];
  defaultLanguage: string;
  inputDir?: string;
  outputDir?: string;
  dynamicLanguage?: string[];
};
export type KeyVars = { [k: string]: string[] };
export type AllKeys = { [key: string]: { [lang: string]: KeyValue } };
export type KeyValue = {
  value: string;
  vars: string[];
  raw: any;
};
