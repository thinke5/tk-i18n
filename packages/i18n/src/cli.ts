#!/usr/bin/env node

import { run } from './index';

// console.log(process.argv);

run();
