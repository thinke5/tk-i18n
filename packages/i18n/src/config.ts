import { cwd } from 'node:process';
import { readJsonFile } from './utils/readJsonFile';
import { Config } from './type';

const measageDir = 'messages';

let _config: Required<Config>;
/** 获取所有配置 */
export async function getConfig() {
  const basePath = cwd();
  if (!_config) {
    const config = await readJsonFile<Config>(basePath, 'i18n.config');

    let {
      language,
      defaultLanguage,
      inputDir = measageDir,
      outputDir = './src/i18n',
      dynamicLanguage = [],
      multipleFilesLanguage = {},
    } = config;
    if (!language) {
      console.error('config language is empty');
      process.exit(2);
    }
    if (!defaultLanguage) {
      defaultLanguage = language[0]; // 默认取第一个语言
    }
    _config = {
      language,
      defaultLanguage,
      inputDir,
      outputDir,
      dynamicLanguage,
      multipleFilesLanguage,
    };
  }

  return _config;
}

/** 是否存在动态增加的语言 */
export async function hasDynamicLanguage() {
  return (await getConfig()).dynamicLanguage.length > 0;
}
/** 获取所有语言，包含动态语言 */
export async function getAllLanguageList() {
  const config = await getConfig();
  return config.language.concat(...config.dynamicLanguage);
}
