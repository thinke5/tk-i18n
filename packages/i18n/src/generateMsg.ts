export function generateMrg(param: { data: object; keys: string[]; langs: string[] }) {
  const { data, keys, langs } = param;
  const result = '';
  keys.forEach((key) => {
    return [
      //
      '/**',
      ' */',
      '/* @__NO_SIDE_EFFECTS__ */',
      `export const ${key} = (param = {}) => {`,

      '}',
    ].join('\n');
  });
  return '';
}
