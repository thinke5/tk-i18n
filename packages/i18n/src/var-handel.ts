import { StrAst, VarHandle } from './type';
import { splitString } from './utils';

/** 如何处理变量，按倒序调用 */
const handles: VarHandle[] = [];
/** 获取所有处理函数 */
export function getVarHandles() {
  return handles;
}
/** 新增处理函数 */
export function pushVarHandle(...v: VarHandle[]) {
  return handles.push(...v);
}
/** 获取合适的处理函数 */
export function getVarHandle(varStr: string) {
  let res: StrAst;
  for (let i = handles.length - 1; i >= 0; i--) {
    const handle = handles[i];
    const v = handle.separate(varStr);
    if (v) {
      res = v;
      break;
    }
  }
  return res!;
}

/** 最基础的变量
 * @example `{name}` => param.name
 */
const baseVarHandel: VarHandle = {
  type: 'var',
  separate: (str) => ({
    type: 'var',
    value: str,
    toString: () => addTemple(`param.${str}`),
    varName: str,
  }),
};

/** 含有处理函数 的变量
 * @description 可以是任意 `js全局函数` 或 `文件内的字段` 或 `方法`
 * @example 1 `{name|..length}` => param.name.length
 * @example 2 `{name|>length(#)}` => length(param.name)
 * @example 3 `{name|..length|>sum(#,2)}` => sum(param.name.length,2)
 */
export const VarHaveHandel: VarHandle = {
  type: 'var-handle',
  separate: (str) => {
    const [name, ...handleStr] = splitString(str, '|');
    if (str.includes('|.') || str.includes('|>')) {
      return {
        type: 'var-handle',
        value: str,
        toString: () => {
          return addTemple(pipeline('param.' + name, handleStr));
        },
        varName: name,
      };
    }
    return null;
  },
};

/** 含有自身文件内 的变量
 * @description 支持处理函数
 * @example 1 `{@key}` => key()
 * @example 2 `{@key|..lenght}` => key().lenght
 * @example 2 `{@key|>sum(#)}` => sum(key())
 */
export const VarThis: VarHandle = {
  type: 'var-this',
  separate: (str) => {
    if (/^\@/.test(str)) {
      const v = str.slice(1);
      const [name, ...handleStr] = splitString(v, '|');
      return {
        type: 'var-this',
        value: str,
        toString: () => {
          // todo param传递
          let res = `${v}()`;

          if (handleStr.length > 0) {
            res = pipeline(`${name}()`, handleStr);
          }

          return addTemple(res);
        },
      };
    }
    return null;
  },
};

pushVarHandle(baseVarHandel, VarHaveHandel, VarThis);

/** 处理管道 */
function pipeline(base: string, handleStr: string[]) {
  let res = base;
  for (const str of handleStr) {
    if (/^\./.test(str)) {
      res += str.slice(1);
    }
    if (/^\>/.test(str)) {
      res = str.slice(1).replace('#', res);
    }
  }

  return res;
}

function addTemple(str: string) {
  return '${' + str + '}';
}
