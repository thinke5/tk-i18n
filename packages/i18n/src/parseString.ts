import { Ast, StrAst } from './type';
import { getVarHandle } from './var-handel';

const reg = /(\{[^}]+\})/g;
const varTag = '$var$';

export function parseString(str: string) {
  const vars: string[] = [];
  const rt: Ast = [];
  str
    .replace(reg, (v) => {
      vars.push(getVarName(v));
      return varTag;
    })
    .split(varTag)
    .filter(Boolean)
    .forEach((txt, i) => {
      rt.push({
        type: 'text',
        value: txt,
        toString() {
          return txt;
        },
      });
      if (vars[i]) {
        rt.push(getVarHandle(vars[i]));
      }
    });

  return rt;
}

/** 移除 { } */
function getVarName(str: string) {
  return str.slice(1, -1);
}
