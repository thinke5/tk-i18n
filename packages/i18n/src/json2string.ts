import { parseString } from './parseString';
import { KeyValue } from './type';

export function handleVarsAny(value: any) {
  let res: KeyValue = { value, vars: [], raw: value };
  if (typeof value === 'string') {
    res = handleString(value);
  }
  if (Array.isArray(value)) {
    res = handleArray(value);
  }
  if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
    res = hanleObject(value);
  }

  return res;
}

function hanleObject(obj: { [key: string]: any }): KeyValue {
  const vars = new Set<string>();
  let values: string[] = [];
  Object.entries(obj).forEach(([key, value]) => {
    const obj = handleVarsAny(value);
    obj.vars.forEach((v) => vars.add(v));

    values.push(`${key}: ${obj.value}`);
  });
  const value = `({${values.join(', ')}})`;
  return { value, vars: Array.from(vars), raw: obj };
}

function handleArray(arr: any[]): KeyValue {
  const vars = new Set<string>();
  let values: string[] = [];
  arr.forEach((str) => {
    const obj = handleVarsAny(str);
    obj.vars.forEach((v) => vars.add(v));

    values.push(obj.value);
  });
  const value = `([${values.join(', ')}])`;
  return { value, vars: Array.from(vars), raw: arr };
}

function handleString(str: string): KeyValue {
  let value = '';
  const vars = new Set<string>();
  parseString(str).forEach((item) => {
    value += item.toString();
    item.varName && vars.add(item.varName);
  });
  value = `\`${value}\``;
  return { value, vars: Array.from(vars), raw: str };
}
