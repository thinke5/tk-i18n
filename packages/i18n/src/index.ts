import { mkdir, readFile, writeFile, copyFile } from 'node:fs/promises';
import * as path from 'node:path';
import { cwd } from 'node:process';
import { handleVarsAny } from './json2string';
import { AllKeys, Config } from './type';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const measageDir = 'messages';

export async function run() {
  console.log('start i18n');

  const basePath = cwd();
  const file = await readFile(path.resolve(basePath, 'i18n.config.json'));
  const config: Config = JSON.parse(file.toString());
  const { language, defaultLanguage, inputDir = measageDir, outputDir = './src/i18n' } = config;

  if (!language) {
    console.error('config language is empty');
    process.exit(2);
  }
  if (!defaultLanguage) {
    console.error('config defaultLanguage is empty');
    process.exit(2);
  }

  const langMsgPath = path.resolve(basePath, outputDir, measageDir);
  await mkdir(langMsgPath, { recursive: true });

  const allKeys: AllKeys = {};
  // cin
  for (const lang of language) {
    const filePath = path.resolve(basePath, inputDir, `${lang}.json`);
    const file = await readFile(filePath).catch((err) => {
      if (err.errno === -2) {
        console.error('!!! file not found: ' + filePath);
      }
    });
    let json: { [key: string]: string } = {};
    if (file) {
      json = JSON.parse(file.toString());
    }
    Object.entries(json).map(([key, value]) => {
      const v = handleVarsAny(value); // 处理字符串
      // 统计数据
      if (allKeys[key]) {
        allKeys[key][lang] = v;
      } else {
        allKeys[key] = { [lang]: v };
      }
    });
  }
  // handle msg
  const langMsgs: { [lang: string]: string } = {};
  let mainMsgs = [
    ...language.map((l) => `import * as ${l} from './messages/${l}.mjs';`),
    "import { currentLanguage } from './runtime.mjs';",
    '',
  ].join('\n');
  Object.entries(allKeys).forEach(([key, value]) => {
    // lang
    language.forEach((lang) => {
      if (langMsgs[lang] === undefined) {
        langMsgs[lang] = '';
      }
      // 生成各个语言文件
      if (value[lang]) {
        langMsgs[lang] += toLangMsg({ key, ...value[lang] });
      } else if (value[defaultLanguage] && lang !== defaultLanguage) {
        langMsgs[lang] += `\n\nexport { ${key} } from "./${defaultLanguage}.mjs"`;
      } else {
        //  默认语言 没有，查找其他第一个非空的
        const v = language.find((l) => value[l])!;
        langMsgs[lang] += toLangMsg({ key, ...value[v] });
      }
    });
    // all
    const allVars = Array.from(new Set(Object.values(value).map((v) => v.vars))).flat();
    mainMsgs += toMainMsg({ key, language, vars: allVars, allKeys, defaultLanguage });
  });
  // 写入 runtime
  readFile(path.resolve(__dirname, './files/runtime.mjs'))
    .then((res) =>
      res
        .toString()
        .replace(`'defaultLanguage'`, JSON.stringify(defaultLanguage))
        .replace(`['defaultLanguage']`, JSON.stringify(language))
    )
    .then((str) => writeFile(path.resolve(langMsgPath, `../runtime.mjs`), str));
  // 写入文件
  writeLangMsg(langMsgPath, langMsgs);
  writeMainMsg(langMsgPath, mainMsgs);
  // end
  console.log('end i18n');
}

function toMainMsg(param: {
  vars: string[];
  key: string;
  language: string[];
  allKeys: AllKeys;
  defaultLanguage: string;
}) {
  const hasVar = param.vars.length > 0;

  return [
    '\n\n/**',
    ...param.language.map(
      (lang) => ` * @lang_${lang} \`\`\`\n * ${JSON.stringify(param.allKeys[param.key][lang]?.raw)}\n * \`\`\``
    ),
    hasVar ? ` * @param {{ ${param.vars.map((v) => v + ': any;').join(' ')} }} param` : ' * @param { {} } [param]',
    ` * @param {{ languageTag?: ${param.language.map((l) => `"${l}"`).join('|')} }} options`,
    ' */',
    '/* @__NO_SIDE_EFFECTS__ */',
    `export const ${param.key} = (param, options = {}) => {`,
    '  const lang = options.languageTag || currentLanguage();',
    param.language.map((l) => `  if (lang === "${l}") return ${l}.${param.key}(${hasVar ? 'param' : ''});`).join('\n'),
    `  return ${param.defaultLanguage}.${param.key}(${hasVar ? 'param' : ''});`,
    '};',
  ]
    .filter(Boolean)
    .join('\n');
}

function toLangMsg(param: { value: string; vars: string[]; key: string }) {
  const hasVar = param.vars?.length > 0;
  return [
    '\n\n',
    '/** ```js',
    ' * ' + JSON.stringify(param.value),
    ' * ```',
    hasVar ? ` * @param {{ ${param.vars.map((v) => v + ': any;').join(' ')} }} param` : '',
    ' */',
    '/* @__NO_SIDE_EFFECTS__ */',
    `export const ${param.key} = (${hasVar ? 'param' : ''}) => ${param.value};`,
  ]
    .filter(Boolean)
    .join('\n');
}
const eslint_disable = '/* eslint-disable */\n';

async function writeMainMsg(_path: string, str: string) {
  writeFile(path.resolve(_path, `../.gitignore`), '*');
  writeFile(path.resolve(_path, `../.prettierignore`), '*');
  ['index.mjs', 'index.d.ts'].forEach((v) =>
    copyFile(path.resolve(__dirname, './files/', v), path.resolve(_path, `../`, v))
  );

  await writeFile(path.resolve(_path, `../message.mjs`), eslint_disable + str);
}

async function writeLangMsg(_path: string, msgs: { [lang: string]: string }) {
  for (const lang in msgs) {
    const str = msgs[lang];
    const msgFilePath = path.resolve(_path, `${lang}.mjs`);
    await writeFile(msgFilePath, eslint_disable + str);
  }
}
