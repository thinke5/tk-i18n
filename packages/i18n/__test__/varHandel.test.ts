import { expect, test } from 'vitest';
import { VarHaveHandel, VarThis } from '../src/var-handel';

const VarHaveHandel_okStr = [
  'namestr|..length',
  'namestr|>length(#)',
  'namestr|..length|.*2',
  'namestr|>length(#)|>sum(#)',
  'namestr|>length(#)|>sum(#,2)',
  'namestr|>length(#)|>sum(#,2)|..length',
  'namestr|>length(#)|>sum(#,2)|..length|>abs(#)',
];

test('VarHaveHandel separate', () => {
  const noStr = ['namestr', '@bfg'];
  noStr.forEach((str) => expect(VarHaveHandel.separate(str)).toBe(null));
  VarHaveHandel_okStr.forEach((str) => expect(VarHaveHandel.separate(str)).toBeTypeOf('object'));
});

test('VarHaveHandel toString', () => {
  VarHaveHandel_okStr.forEach((str) => {
    expect(VarHaveHandel.separate(str)!.toString()).toMatchSnapshot(str);
  });
});

const VarThis_okStr = [
  '@name',
  '@key',
  '@key|..lenght',
  '@key|>sum(#)',
  '@key|>sum(#)|..lenght',
  '@key|..lenght|>sum(#)',
];
const VarThis_noStr = ['name', 'ke@y'];

test('VarThis base', () => {
  VarThis_noStr.forEach((str) => expect(VarHaveHandel.separate(str)).toBe(null));

  VarThis_okStr.forEach((str) => {
    expect(VarThis.separate(str)!.toString()).toMatchSnapshot(str);
  });
});
