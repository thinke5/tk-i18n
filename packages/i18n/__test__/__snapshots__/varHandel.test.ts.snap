// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VarHaveHandel toString > namestr|..length 1`] = `"\${param.namestr.length}"`;

exports[`VarHaveHandel toString > namestr|..length|.*2 1`] = `"\${param.namestr.length*2}"`;

exports[`VarHaveHandel toString > namestr|>length(#) 1`] = `"\${length(param.namestr)}"`;

exports[`VarHaveHandel toString > namestr|>length(#)|>sum(#) 1`] = `"\${sum(length(param.namestr))}"`;

exports[`VarHaveHandel toString > namestr|>length(#)|>sum(#,2) 1`] = `"\${sum(length(param.namestr),2)}"`;

exports[`VarHaveHandel toString > namestr|>length(#)|>sum(#,2)|..length 1`] = `"\${sum(length(param.namestr),2).length}"`;

exports[`VarHaveHandel toString > namestr|>length(#)|>sum(#,2)|..length|>abs(#) 1`] = `"\${abs(sum(length(param.namestr),2).length)}"`;

exports[`VarThis base > @key 1`] = `"\${key()}"`;

exports[`VarThis base > @key|..lenght 1`] = `"\${key().lenght}"`;

exports[`VarThis base > @key|..lenght|>sum(#) 1`] = `"\${sum(key().lenght)}"`;

exports[`VarThis base > @key|>sum(#) 1`] = `"\${sum(key())}"`;

exports[`VarThis base > @key|>sum(#)|..lenght 1`] = `"\${sum(key()).lenght}"`;

exports[`VarThis base > @name 1`] = `"\${name()}"`;
